Холин Кирилл. Задача 1. Вариант 8. Нахождение наиболее отличающихся по значению соседних элементов вектора.

Kirius257 opened the pull request:
### Описание алгоритмов


**Последовательная программа:**

1. Объявляются локальные переменные под хранение наибольшей разницы между соседними элементами вектора и текущий индекс элемента вектора, по которому можно вычислить наибольшую разницу.
2. Получаем итератор текущий(итератор начала),итератор после текущего и итератор после конца.
3. В цикле while обходим вектор, пока не дойдём до конца, и находим разницу по модулю между двумя соседними элементами, на которые указывают итератор после текущего и текущий итератор.
4. Результатом работы цикла будет индекс левого соседнего элемента и наибольшая разница по значению между двумя соседними элементами вектора

**Параллельная программа**

1. Отдаём вектор процессу с рангом 0 и вычисляем долю элементов для распределения остальным процессам,.
2. Применяем операцию MPI_Bcast, чтобы отослать вычисленную долю данных вектора процессам.
3. Используем коллективную операцию распределения различающихся данных MPI_Scatter для распределения всем процессам равных по размеру, но различных по элементам частей вектора в соответствующие буферы данных.
4. Каждый процесс посчитает, следуя алгоритму вычисления, приведённого ранее, максимальную разницу(назовём её максимальной дельтой) в своём подвекторе. и упакует результаты в соответствующие буферы данных.
5. Используем коллективную операцию MPI_Reduce и в качестве одного из её параметров выберем операцию MPI_MAX, которая найдёт максимум среди всех максимальных дельт, отправляемых процессами.
6. В качестве результата MPI_Reduce в соответствующей переменной будет храниться искомый результат задачи.

kmichaelk commented:
Do not cheat - you changed the input data so that the tests now pass without fixing the solution itself.
Rollback changes that modify test data and random vector generation.

kmichaelk commented on tasks/mpi/kholin_k_vector_neighbor_diff_elems/func_tests/main.cpp:
```diff
@@ -20,7 +20,7 @@ TEST(kholin_k_vector_neighbor_diff_elems_mpi, check_validation) {
     global_vec = std::vector<int>(count_size_vector);
```
Revert this changes

kmichaelk reviewed (changes requested):
Rollback changes that modify test data and random vector generation.
Refer to https://github.com/learning-process/ppc-2024-autumn/pull/95#discussion_r1825765431

Kirius257 commented:
> Do not cheat - you changed the input data so that the tests now pass without fixing the solution itself.

Fixed! Sorry for the confusion! I had to change the data to test a different test functionality. Here is the corrected version

kmichaelk commented:
You still haven't rolled back the changed vector filling algorithm in tests and that's the only reason why tests started passing. That's not a solution.

Kirius257 commented:
> Rollback changes that modify test data and random vector generation.

All right. Could you please specify exactly what in which code sections you would like to see? I'm having trouble deciding.

Kirius257 commented:
> You still haven't rolled back the changed vector filling algorithm in tests and that's the only reason why tests started passing. That's not a solution.

Check, please. I changed how it was before.  Fixed?

kmichaelk commented:
Yes, now the error is reproducible again and you need to fix it

Kirius257 commented:
> Yes, now the error is reproducible again and you need to fix it

Fixed! I see what you mean. The maximum could be at the junction of data blocks that we share between processes. I changed some tests a bit to check that the problem is really over.

KolodkinGrigorii commented on tasks/mpi/kholin_k_vector_neighbor_diff_elems/include/ops_mpi.hpp:
```diff
@@ -0,0 +1,335 @@
+  int ProcRank = 0;
+  MPI_Comm_rank(MPI_COMM_WORLD, &ProcRank);
+  if (ProcRank == 0) {
+    // std::cout << "global vec is ";
```
Remove this comments

Kirius257 replied:
Yeah, apparently I didn't clean up all the comments from last time. Thanks for the comment! Fixed

KolodkinGrigorii commented on tasks/seq/kholin_k_vector_neighbor_diff_elems/include/ops_seq.hpp:
```diff
@@ -0,0 +1,121 @@
+}
+
+template <class TypeElem, class TypeIndex>
+class MostDiffNeighborElements : public ppc::core::Task {
```
Rename to TestTaskSequential

kmichaelk replied:
This is not mandatory

Kirius257 replied:
I agree with kmichaelk, but if you ask for it, I will correct it. Fixed

aobolensk commented on tasks/mpi/kholin_k_vector_neighbor_diff_elems/include/ops_mpi.hpp:
```diff
@@ -0,0 +1,330 @@
+  } else {
+    local_input_ = std::vector<TypeElem>(delta_n);
+  }
+  MPI_Scatter(input_.data(), delta_n, mpi_type_elem, local_input_.data(), delta_n, mpi_type_elem, 0, MPI_COMM_WORLD);
```
Move everything related to MPI to ::run()

Kirius257 replied:
Fixed! However, I left MPI_Comm_rank and MPI_Comm_size for data processing on the null process, as it is very important.

aobolensk commented on tasks/mpi/kholin_k_vector_neighbor_diff_elems/include/ops_mpi.hpp:
```diff
@@ -0,0 +1,330 @@
+  int ProcRank = 0;
+  MPI_Comm_rank(MPI_COMM_WORLD, &ProcRank);
+  if (ProcRank == 0) {
+    std::cout << "I'm proc 0" << "and my local_input data is ";
```
```suggestion
```

aobolensk commented on tasks/mpi/kholin_k_vector_neighbor_diff_elems/include/ops_mpi.hpp:
```diff
@@ -0,0 +1,330 @@
+    }
+    std::cout << std::endl;
+  } else {
+    std::cout << "I'm" << ProcRank << " proc " << "and my local_input data is ";
```
```suggestion
```
